packages/classes/src/PayloadArray.js

```diff
@@ -0,0 +1,98 @@
+import Payload from './Payload'
+
+/**
+ * PayloadArray stores a series of Payloads in a data structure that allows for easy filtering
+ * and lookup by position and type.
+ */
+class PayloadArray {
+	payloads = []
+
+	constructor(...payloads) {
+		[...payloads].forEach((payload) => this.insert(payload))
+	}
+
+	get length() { return this.payloads.length }
+
+	/**
+	 * Insert a Payload into the PayloadArray while maintaining position order.
+	 *
+	 * @param  {Payload} payload The Payload to be inserted.
+	 * @return {PayloadArray}    This PayloadArray to allow for chaining.
+	 */
+	insert = (payload) => {
+		if (!(payload instanceof Payload)) {
+			throw new Error('Attempt to insert a non-Payload to PayloadArray')
+		}
+		const index = this.indexOfPosition(payload.position)
+		this.payloads.splice(index, 0, payload)
+		return this
+	}
+
+	/**
+	 * Find the index of the first Payload whose position is greater than or equal to this one.
+	 *
+	 * NOTE: if no qualifying payloads exist, this will return the length of the PayloadArray
```

Comment:
```suggestion
	 * If no qualifying payloads exist, this will return the length of the PayloadArray since the
	 * implied position for the payload is the end of the array.
```